### PR TITLE
Lower const params with a bad id

### PR DESCRIPTION
**Pull request by HKalbasi**
cc #7434 

This PR adds an `InTypeConstId` which is a `DefWithBodyId` and lower const generic parameters into bodies using it, and evaluate them with the mir interpreter. I think this is the last unimplemented const generic feature relative to rustc stable.

But there is a problem: The id used in the `InTypeConstId` is the raw `FileAstId`, which changes frequently. So these ids and their bodies will be invalidated very frequently, which is bad for incremental analysis.

Due this problem, I disabled lowering for local crates (in library crate the id is stable since files won't be changed). This might be overreacting (const generic expressions are usually small, maybe it would be better enabled with bad performance than disabled) but it makes motivation for doing it in the correct way, and it splits the potential panic and breakages that usually comes with const generic PRs in two steps.

Other than the id, I think (at least I hope) other parts are in the right direction.

### CODE REVIEW

File: crates/hir-def/src/hir/type_ref.rs [hunk 1]
```diff
@@ -380,73 +381,84 @@ impl TypeBound {
 }
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub enum ConstRefOrPath {
-    Scalar(ConstRef),
+pub enum ConstRef {
+    Scalar(ConcreteConstRef),
     Path(Name),
+    Complex(InFile<FileAstId<ast::Expr>>),
```
Comment (Veykril) [inline crates/hir-def/src/hir/type_ref.rs:387]:
```suggestion
    Complex(AstId<ast::Expr>),
```
we have a convenient alias for that

File: crates/hir-def/src/hir/type_ref.rs [hunk 2]
```diff
@@ -380,73 +381,84 @@ impl TypeBound {
...
+                }
+            }
         }
     }
 }
 
 /// A concrete constant value
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub enum ConstRef {
+pub enum ConcreteConstRef {
```
Comment (Veykril) [inline crates/hir-def/src/hir/type_ref.rs:461]: I wonder if `LiteralConstRef` would be more descriptive of a name? Though I guess that doesn't fit with the unknown variant rn

File: crates/hir-def/src/lib.rs [hunk 3]
```diff
@@ -476,20 +476,54 @@ impl_from!(
...
+        }
+    }
+}
+
+impl_from!(ModuleId, DefWithBodyId(FunctionId, ConstId, StaticId), GenericDefId(AdtId, TypeAliasId, ImplId) for TypeOwnerId);
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+pub struct InTypeConstId(InternId);
+type InTypeConstLoc =
+    (hir_expand::InFile<hir_expand::ast_id_map::FileAstId<ast::Expr>>, TypeOwnerId);
```
Comment (Veykril) [inline crates/hir-def/src/lib.rs:505]:
```suggestion
    (AstId<ast::Expr>, TypeOwnerId);
```

File: crates/hir-def/src/lib.rs [hunk 4]
```diff
@@ -476,20 +476,54 @@ impl_from!(
     for ModuleDefId
 );
 
-// FIXME: make this a DefWithBodyId
```
Comment (Veykril) [inline crates/hir-def/src/lib.rs:479]: Can we give this a comment describing what construct this is? I am not quite sure what an anon const represents personally

File: crates/hir-expand/src/ast_id_map.rs [hunk 5]
```diff
@@ -98,6 +98,7 @@ impl AstIdMap {
                 || ast::Variant::can_cast(kind)
                 || ast::RecordField::can_cast(kind)
                 || ast::TupleField::can_cast(kind)
+                || ast::Expr::can_cast(kind)
```
Comment (Veykril) [inline crates/hir-expand/src/ast_id_map.rs:101]: This will most likely have a noticable memory impact.

**Comment by Veykril** [comment]
I think before we try to tackle this issue we should take a step back and consider the possible ways to implement it with advantages and disadvantages in mind (maybe as a zulip discussion). This looks like it will have some architecture implications so it feels like we wanna get this right (or close to right) in an initial implementation. If I recall, this approach here is a different from the one that was mentioned in the issue itself right?

**Review by flodiebold — COMMENTED** [review]
some drive-by comments

File: crates/hir-def/src/lib.rs [hunk 6]
```diff
@@ -476,20 +476,55 @@ impl_from!(
...
 #[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
 pub struct AnonymousConstId(InternId);
 impl_intern_key!(AnonymousConstId);
 
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+pub enum TypeOwnerId {
+    ModuleId(ModuleId),
+    DefWithBodyId(DefWithBodyId),
+    GenericDefId(GenericDefId),
+}
```
[the remark this answers is not in the document]
Comment (flodiebold) [reply crates/hir-def/src/lib.rs:490]: Hmm, I guess I was somehow expecting types inside bodies to be handled differently :thinking: Flattening it probably makes sense...

**Pull request merged**